test/unit/gateways/hps_test.rb

```diff
@@ -195,6 +197,229 @@ def test_transcript_scrubbing
+      verification_value: nil,
+      source: :apple_pay
+    )
+    assert response = @gateway.purchase(@amount, credit_card, @options)
+    assert_failure response
+    assert_equal 'The card was declined.', response.message
+  end
+
+  def test_successful_auth_with_apple_pay_raw_cryptogram_with_eci
+    @gateway.expects(:ssl_post).returns(successful_authorize_response)
+
+    credit_card = network_tokenization_credit_card('4242424242424242',
+      payment_cryptogram: 'EHuWW9PiBkWvqE5juRwDzAUFBAk=',
+      verification_value: nil,
+      eci: '05',
+      source: :apple_pay
+    )
+    assert response = @gateway.authorize(@amount, credit_card, @options)
+    assert_success response
+    assert_equal 'Success', response.message
+  end
+
+  def test_failed_auth_with_apple_pay_raw_cryptogram_with_eci
+    @gateway.expects(:ssl_post).returns(failed_authorize_response_decline)
+
+    credit_card = network_tokenization_credit_card('4242424242424242',
+      payment_cryptogram: 'EHuWW9PiBkWvqE5juRwDzAUFBAk=',
+      verification_value: nil,
+      eci: '05',
+      source: :apple_pay
+    )
+    assert response = @gateway.authorize(@amount, credit_card, @options)
+    assert_failure response
+    assert_equal 'The card was declined.', response.message
+  end
+
+  def test_successful_auth_with_apple_pay_raw_cryptogram_without_eci
+    @gateway.expects(:ssl_post).returns(successful_authorize_response)
+
+    credit_card = network_tokenization_credit_card('4242424242424242',
+      payment_cryptogram: 'EHuWW9PiBkWvqE5juRwDzAUFBAk=',
+      verification_value: nil,
+      source: :apple_pay
+    )
+    assert response = @gateway.authorize(@amount, credit_card, @options)
+    assert_success response
+    assert_equal 'Success', response.message
+  end
+
+  def test_failed_auth_with_apple_pay_raw_cryptogram_without_eci
+    @gateway.expects(:ssl_post).returns(failed_authorize_response_decline)
+
+    credit_card = network_tokenization_credit_card('4242424242424242',
+      payment_cryptogram: 'EHuWW9PiBkWvqE5juRwDzAUFBAk=',
+      verification_value: nil,
+      source: :apple_pay
+    )
+    assert response = @gateway.authorize(@amount, credit_card, @options)
+    assert_failure response
+    assert_equal 'The card was declined.', response.message
+  end
+
+  def test_three_d_secure_visa
+    @credit_card.number = '4012002000060016'
+    @credit_card.brand = 'visa'
+
+    options = {
+      :three_d_secure => {
+        :cavv => 'EHuWW9PiBkWvqE5juRwDzAUFBAk=',
+        :eci => '05',
+        :xid => 'TTBCSkVTa1ZpbDI1bjRxbGk5ODE='
+      }
+    }
+
+    response = stub_comms(@gateway, :ssl_request) do
+      @gateway.purchase(@amount, @credit_card, options)
+    end.check_request do |method, endpoint, data, headers|
+      assert_match(/<hps:SecureECommerce>(.*)<\/hps:SecureECommerce>/, data)
+      assert_match(/<hps:PaymentDataSource>Visa 3DSecure<\/hps:PaymentDataSource>/, data)
+    end.respond_with(successful_charge_response)
```

Comment:
I would expand the assertions for the new 3DS tests to add something like the following (which verifies the ECI leading-zero-stripping is happening as expected and that the other values from the 3DS options are correctly input into the resultant XML request document):

```
      assert_match(/<hps:PaymentData>#{options[:three_d_secure][:cavv]}<\/hps:PaymentData>/, data)
      assert_match(/<hps:ECommerceIndicator>5<\/hps:ECommerceIndicator>/, data)
      assert_match(/<hps:XID>#{options[:three_d_secure][:xid]}<\/hps:XID>/, data)
```

Reply:
Sounds a good improvement to add coverage around ECI zero stripping and 3DS. @slogsdon Could we have that updated pls. Looks like we are almost there.

Reply:
Done via cc218668e61a6c53ab0f415de156455e20411e74.